### PR TITLE
Rebase #2431 against proposed/3.0.0

### PR DESCRIPTION
**Pull request by alexwlchan**
This is really quite unexciting: #2431 had a tentative +1 from @Lukasa, pending a rebase and retargeting against 3.0.0, but it’s been sat unloved for nearly eight months. 😢 

This patch does the required rebase and targets the 3.0.0 branch.

### CODE REVIEW

File: tests/test_requests.py [hunk 1]
```diff
@@ -1203,6 +1203,46 @@ def test_request_and_response_are_pickleable(self, httpbin):
...
+            '''Fake difficult data.'''
+            for chunk in mock_chunks:
+                yield chunk
+
+        r = requests.Response()
+        r._content_consumed = True
+        r.iter_content = mock_iter_content
+
+        assert list(r.iter_lines(delimiter='\r\n')) == \
+            ''.join(mock_chunks).split('\r\n')
```
Comment (Lukasa) [inline tests/test_requests.py]: I'd like these asserts cleaned up if at all possible, they're kinda complex. Let's break some stuff up so the assert is somewhat self-describing.

File: requests/models.py [hunk 2]
```diff
@@ -753,23 +753,36 @@ def iter_lines(self, chunk_size=ITER_CHUNK_SIZE, decode_unicode=None, delimiter=
...
-            if lines and lines[-1] and chunk and lines[-1][-1] == chunk[-1]:
+            # The split(delimiter) will always end with whatever remains past
+            # the delimiter ('' if nothing more).  However splitlines() will
+            # not end with a '' if the final text is a line delimiter.
+
+            # Therefore, if we're in delimiter mode, always pop the final
+            # item to prepend to the next chunk. However, only do this for
+            # non-delimiter mode if the chunk does not match the end of the
+            # last line.
+            if delimiter or (lines[-1] and lines[-1][-1] == chunk[-1]):
```
Comment (Lukasa) [inline requests/models.py]: Can we break out the parenthetical here and give it a name that hopefully describes what we're actually checking?

**Comment by ianjepperson** [comment]
Woah, zombie PR back from the dead!  I haven't worked in Python for over a year now, but I'll see if I can address the suggestions this weekend.

**Comment by alexwlchan** [comment]
I had a go at addressing the review markups, but I feel like more testing is required.

I feel like there’s enough subtle behaviour here that we’d be better off with a whole stack of tests (probably using `@pytest.mark.parametrize`) that tests a whole variety of cases here. Which I may or may not get around to writing sometime this week. At a minimum, the cases described in [this comment](https://github.com/kennethreitz/requests/pull/2431#issuecomment-72333964).

File: requests/models.py [hunk 3]
```diff
@@ -737,42 +737,70 @@ def generate():
...
             # check encoding value here, don't wait for the generator to be
             # consumed before raising an exception
             codecs.lookup(self.encoding)
-            
+
             chunks = stream_decode_response_unicode(chunks, self)
 
         return chunks
 
+    import q
```
Comment (sigmavirus24) [inline requests/models.py]: What is `q`?

Comment (alexwlchan) [reply requests/models.py]: Oh damn, somebody checked in his debugging code without checking the diff, didn't he?

My oopsie, will fix.

([q](https://pypi.python.org/pypi/q/2.6) is a module for quick and dirty debugging that has no business in a production codebase)

**Comment by Lukasa** [comment]
Yeah, I think more testing here would be fab. Not sure how we're going to get it though, especially if @alexwlchan doesn't have time to do it.

**Comment by alexwlchan** [comment]
I’ve stuck a note in my todo list to come back and write more tests here. I’ve built up a bit of a backlog of unfinished PRs, but hopefully I’ll find time to get back to this over Christmas.

**Comment by kennethreitz** [comment]
Closing this out due to lack of activity. Feel free to re-open once you get a chance!

**Pull request closed without merging**